### PR TITLE
refactor(amazonq) refactor ProjectContextProvider.kt and add tests

### PR DESCRIPTION
**Pull request by Will-ShaoHua**
cherry pick from #4976 
- remove repeated code
- make message toward lsp type safe
- add test coverage

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/project/ProjectContextProvider.kt [hunk 1]
```diff
@@ -53,25 +55,12 @@ class ProjectContextProvider(val project: Project, private val encoderServer: En
             }
         }
     }
-    data class IndexRequestPayload(
-        val filePaths: List<String>,
-        val projectRoot: String,
-        val refresh: Boolean,
-    )
```
Comment (Will-ShaoHua) [inline plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/project/ProjectContextProvider.kt:60]: move to https://github.com/aws/aws-toolkit-jetbrains/pull/4988/files#diff-00c500185316c5fcdd5fdf6196f00012ecfd3abb9fa9b0851c773d92a2e6e796R32-R36

File: plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/project/ProjectContextProvider.kt [hunk 2]
```diff
@@ -53,25 +55,12 @@ class ProjectContextProvider(val project: Project, private val encoderServer: En
...
         val fileSize: Int,
     )
 
-    data class QueryRequestPayload(
-        val query: String,
-    )
-
-    data class UpdateIndexRequestPayload(
-        val filePath: String,
-    )
```
Comment (Will-ShaoHua) [inline plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/project/ProjectContextProvider.kt:73]: move to https://github.com/aws/aws-toolkit-jetbrains/pull/4988/files#diff-00c500185316c5fcdd5fdf6196f00012ecfd3abb9fa9b0851c773d92a2e6e796R38-R44

File: plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonq/workspace/context/ProjectContextProviderTest.kt [hunk 3]
```diff
@@ -57,4 +243,50 @@ class ProjectContextProviderTest {
         }
         verify(encoderServer, times(1)).encrypt(any())
     }
+
+    private fun createMockServer() = WireMockRule(wireMockConfig().dynamicPort())
 }
+
+val validQueryChatResponse = """
```
Comment (rli) [inline plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonq/workspace/context/ProjectContextProviderTest.kt:250]: can you reformat the JSON blobs
```suggestion
// language=JSON
val validQueryChatResponse = """
```

File: plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/project/ProjectContextProvider.kt [hunk 4]
```diff
@@ -346,6 +286,34 @@ class ProjectContextProvider(val project: Project, private val encoderServer: En
...
+    private fun encryptRequest(r: LspRequest): String {
+        val payloadJson = mapper.writeValueAsString(r)
+        return encoderServer.encrypt(payloadJson)
+    }
+
+    private fun sendMsgToLsp(msgType: LspMessage, request: String?): LspResponse {
+        logger.info { "sending message: ${msgType.endpoint} to lsp on port ${encoderServer.port}" }
+        val url = URL("http://localhost:${encoderServer.port}/${msgType.endpoint}")
+
+        return with(url.openConnection() as HttpURLConnection) {
```
Comment (rli) [inline plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/project/ProjectContextProvider.kt:298]: consider using a proper http client

**Review by rli — APPROVED** [review]

File: plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/project/ProjectContextProvider.kt (hunk 4 above)
Comment (Will-ShaoHua) [reply plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/project/ProjectContextProvider.kt:298]: ack

File: plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonq/workspace/context/ProjectContextProviderTest.kt (hunk 3 above)
Comment (Will-ShaoHua) [reply plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonq/workspace/context/ProjectContextProviderTest.kt:250]: @rli what does `// language=JSON` mean here

Comment (Will-ShaoHua) [reply plugins/amazonq/chat/jetbrains-community/tst/software/aws/toolkits/jetbrains/services/amazonq/workspace/context/ProjectContextProviderTest.kt:250]: isee, it tells IDE to treat it as JSON instead plain string

File: plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/project/ProjectContextProvider.kt (hunk 4 above)
Comment (Will-ShaoHua) [reply plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/project/ProjectContextProvider.kt:298]: https://docs.oracle.com/en/java/javase/17/docs/api/java.net.http/java/net/http/package-summary.html

**Review by leigaol — APPROVED** [review]

File: plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/project/ProjectContextProvider.kt [hunk 5]
```diff
@@ -172,34 +151,35 @@ class ProjectContextProvider(val project: Project, private val encoderServer: En
...
+            parsedResponse
         } catch (e: Exception) {
             logger.warn { "error parsing query response ${e.message}" }
-            return emptyList()
+            null
         }
     }
 
+    fun updateIndex(filePath: String) {
+        if (!isIndexComplete.get()) return
```
Comment (Will-ShaoHua) [inline plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/project/ProjectContextProvider.kt:178]: should remove this line.

File: plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/project/ProjectContextProvider.kt [hunk 6]
```diff
@@ -130,37 +119,27 @@ class ProjectContextProvider(val project: Project, private val encoderServer: En
...
-        logger.info { "project context initialize response code: ${connection.responseCode} for ${project.name}" }
-        return connection.responseCode == 200
+        val request = encoderServer.getEncryptionRequest()
+        val response = sendMsgToLsp(LspMessage.Initialize, request)
+        return response.responseCode == 200
     }
 
     fun index(): Boolean {
-        logger.info { "project context: indexing ${project.name} on port ${encoderServer.port}" }
+        val projectRoot = project.guessProjectDir()?.path ?: return false
```
Comment (leigaol) [inline plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/project/ProjectContextProvider.kt:128]: regression. Please fix ASAP

**Pull request merged**